File: 프로그래머스/lv2/49993. 스킬트리/스킬트리.js
function solution(skill, skill_trees) {
    const divideSkill = skill.split('');
    const sliceSkillTree = [];
    
    for(let i = 1; i <= divideSkill.length; i++) {
        sliceSkillTree.push(divideSkill.slice(0, i).join(''));
    }
    
    let count = 0;
    
    skill_trees.forEach(skills => {
        const stack = skills.split('').filter(skill => divideSkill.includes(skill));
        
        if(stack.length === 0) count++;
        else {
            if(sliceSkillTree.includes(stack.join(''))) count++;
        }
    })
    return count;
}

/*
const divideSkill = skill.split('');
    const restSkillTrees = skill_trees.map((skillTree) => skillTree.split('').filter(skill => {
        divideSkill.includes(skill);
    }));
    
    const sliceSkillTree = [];
    
    for(let i = 1; i <= divideSkill.length; i++) {
        sliceSkillTree.push(divideSkill.slice(0, i).join(''));
    }
    
    return restSkillTrees.filter(skillTree => sliceSkillTree.includes(skillTree.join(''))).length
*/
/*
    결국 저 스킬의 순서가 문자열에서 일치하는지 보면 되는거임. 
    나머지 문자는 다 빼고 봐도될거같은데,
    
    나머지 문자를 다 빼고 join해서 그게 
    아아아 일치랑 포함도 있네
    
    (str.includes('Java'));
*/

File: 프로그래머스/2/49993. 스킬트리/스킬트리.py
def delete_(s, skill):
    sl = list(s)
    res = []
    for i in range(len(sl)):
        try:
            index = skill.index(sl[i])
            res.append(sl[i])
        except:
            continue
    return res

def solution(skill, skill_trees):
    skill_s = set()
    for i in range(1, len(skill) + 1):
        sub_ = skill[:i]
        skill_s.add(sub_)
        
    c = 0
    for i in range(len(skill_trees)):
        clean_skill_tree = ''.join(delete_(skill_trees[i], skill))
        
        if len(clean_skill_tree) == 0:
            c += 1
            
        elif clean_skill_tree in skill_s:
            c += 1
    
    return c
            
    
    
